testing: introduce RobotFramework tests

smlng opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This introduces the RobotFramework for writing and running RIOT test, it ships with some adapted tests. Integration into the build system is provided by a new `make` target: `robot-test`. Also adapted
the static tests, just call `make robot-static-test`.

### Testing procedure

Simply build and execute a robot test, e.g. by calling:

```
make -C tests/bench_sizeof_coretypes all robot-test
```

tested on various boards such as `arduino-mega2560`, `pba-d-01-kw2x`, `samr21-xpro` and `native`.

### Issues/PRs references

currently depends on #10038 (maybe adapted, see discussion there)

MrKevinWeiss commented:
Can you somehow add a tag into the output xml that contains the board, test, and commit hash?

smlng commented:
in theory you "know" all this because you invoke it like `BOARD=something make -C tests/test-app flash robot-test`, but I guess you would like to see it in the HTML output, right?

smlng commented:
@MrKevinWeiss I removed the `robot-open` bc it doesn't work on Linux as expected. I also added some logging output including all the stuff you wanted.

MrKevinWeiss commented:
More useful in the xml output.  If it is possible that it just makes everything easier and more modular.  Lets say we feed the output.xml to a data agitator, then we don't need anything special to pass or parse, just the xml.

smlng commented:
@MrKevinWeiss please try again I added the output you wanted

smlng commented:
its now part of the suite setup

MrKevinWeiss commented:
> its now part of the suite setup

Perfect, should be very easy to integrate!

tcschmidt commented:
@cladmi are you just saying that you don't like the robot way of describing tests, i.e., the framework? Or do you propose to express tests differently/more clearly in the robot keywords?

cladmi commented:
@tcschmidt I would say the way of describing tests with the framework. I do not really know how it could be expressed differently anyway.

If I had no idea what the [original](https://github.com/RIOT-OS/RIOT/blob/104de7b6211e93b74a6cf7399e5019e224810685/tests/sizeof_tcb/tests/01-run.py) (now deleted) test did, it would take me 5 real minutes to understand this https://github.com/RIOT-OS/RIOT/blob/b92ab767f91ed8ae1807d5be7050f38f4ff23b0a/tests/bench_sizeof_coretypes/tests/bench_sizeof_coretypes.robot and way longer to write it.

Also after understanding the concept and keywords I still find the python version clearer.

Could the `original` version have a comment on the first detection for if it is 16 or 32 bit ? Yes it could have. But it is not a language issue.


If a documentation is needed before an assert function call I think it is a bad sign

https://github.com/RIOT-OS/RIOT/blob/b92ab767f91ed8ae1807d5be7050f38f4ff23b0a/tests/xtimer_hang/tests/xtimer_hang.robot#L14-L15

tcschmidt commented:
@cladmi I understand your concern. On the other hand, we are comparing apples with oranges: In the specific example, the python test is a very static, fixed example, while the robot test works dynamically for changing configurations.
All in all, python is a generic scripting language that most people are familiar with. So easy to approach, but things must be handcrafted individually. The robot framework is specifically designed for testing and solves many generic problems ... but is specific. However, robot keywords may be chosen according to the domain context and 'intuition'. 
It should also be noted that robot is well established with a large community of users and developers ... so nothing out of a weird niche.

kaspar030 commented:
> it ships with some adapted tests.

Did you try modeling our release specs with the robot framework?

I'm curious how compilation/flashing/handling of multiple nodes would look like. I remember finding it not trivial to implement a generic way to e.g., ping the link-local address of the 802154 interface of node A from node B without knowing it beforehand, or testing node A with examples/gnrc_networking compiled with "GNRC_SETTING_FOO=whiskybar", node B with "GNRC_SETTING_FOO=milkybar".

tcschmidt commented:
Now that the issue #10241 is out, the discussion about the general beauty and usefulness of Robot FW as well as its customization should go there.
This PR is currently only contributing the hook for the (urgently needed) HIL testing suite, which is done with Robot for simplicity. If the discussion in #10241 will lead to a different testing tool, we can switch with the HIL testing later.
@cladmi please review the technical appropriateness here, only and take further aspects to #10241.

kaspar030 commented:
The current build system just executes every executable in the tests/ folder.
Can the robot scripts be executed like that? Why is a seperate make target necessary?

smlng commented:
> The current build system just executes every executable in the tests/ folder.
> Can the robot scripts be executed like that? Why is a separate make target necessary?

##TL;DR 

no it is not possible to make the robot scripts executable.

----

## Longer answer

1.) making the robot test scripts executable will not work, as it is necessary to run them either with `robot` binary or `python -m robot.run`. Further, several parameters are necessary, e.g. where to find libraries and resource files, but also to set output directory and so on. 

2.) the extra make target is not required and the tests could be integrated into the current `make test` target ... that would even be possible without making them executable. However, the idea was to make this first introduction of RF tests *non-intrusive* which also means not to interfere with the testing currently done by Murdock. So it is more a precaution for now and make robot tests available to users without breaking existing stuff. If this works out and we are happy with RF, we can integrate with `make test` and remove `robot-test` target.

miri64 commented:
> 1.) making the robot test scripts executable will not work, as it is necessary to run them either with `robot` binary or `python -m robot.run`. Further, several parameters are necessary, e.g. where to find libraries and resource files, but also to set output directory and so on. 

Contrary to what you probably wanted to say this sounds to me that you *can* make them executable, as they are just scripts to the robot binary / robot module. Have you tried setting a she-bang in line one and setting the `x` flag?

kb2ma commented:
Thanks to @smlng  and @MrKevinWeiss for this work. In the interest of finding a solution, I posted a WIP PR [#86](https://github.com/RIOT-OS/Release-Specs/pull/86) in Release-Specs to use pytest to run the CoAP specs. I think pytest would provide an evolutionary path forward to help with test infrastructure, reuse, and automation while also meeting the objections raised in this issue.

MrKevinWeiss commented:
@cladmi Any update on the blocking comment?  It seems some progress has been made on the RFC.

miri64 commented on dist/robotframework/res/riot_util.keywords.txt:
```diff
@@ -0,0 +1,62 @@
+    ${count} =          Get Number of Lines Matching Pattern  ${pattern}
+    Should Be Equal As Integers  ${count}  ${number}
+
+Verify Lines Matching Pattern Greater
```
As stated elsewhere (can't find it right now): I would prefer if don't re-implementing `pexpect` in RobotFramework keywords. As far as I understand it should be possible to have a thin wrapper around `pexpect` with RobotFramework. This would give us a far more stable and future proof implementation.

smlng commented:
superseded by #10843
